File: services/agent-api/app/schemas/resident_presence.py
"""Realtime compatibility resident presence and hard-stop schemas."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class CompanionResidentStatusEventRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    companion_id: uuid.UUID
    realtime_session_id: uuid.UUID | None = None
    status_type: str
    status_source: str
    interruption_level: str
    allows_unsolicited_presence: bool = False
    presence_summary: str | None = None
    policy_snapshot_json: dict[str, Any] = Field(default_factory=dict)
    occurred_at: datetime | None = None

    model_config = {"from_attributes": True}


class CompanionPresenceBudgetRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    companion_id: uuid.UUID
    budget_scope: str
    budget_status: str
    enforcement_policy: str
    max_presence_minutes: int
    used_presence_minutes: int
    max_interruptions: int
    used_interruptions: int
    window_starts_at: datetime | None = None
    window_ends_at: datetime | None = None
    budget_policy_json: dict[str, Any] = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class CoPresenceInvitationRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    co_presence_session_id: uuid.UUID | None = None
    realtime_session_id: uuid.UUID | None = None
    inviter_companion_id: uuid.UUID | None = None
    target_companion_id: uuid.UUID
    invitation_status: str
    invitation_source: str
    requires_user_approval: bool = True
    auto_join_allowed: bool = False
    memory_candidate_allowed: bool = False
    invitation_reason: str | None = None
    policy_snapshot_json: dict[str, Any] = Field(default_factory=dict)
    expires_at: datetime | None = None

    model_config = {"from_attributes": True}


class QuietHourSettingRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    companion_id: uuid.UUID | None = None
    quiet_status: str
    quiet_policy: str
    day_of_week: int | None = None
    start_minute: int
    end_minute: int
    timezone: str
    allows_emergency_override: bool = False
    policy_json: dict[str, Any] = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class FocusModeEventRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    companion_id: uuid.UUID | None = None
    realtime_session_id: uuid.UUID | None = None
    focus_status: str
    focus_scope: str
    suppress_presence: bool = True
    suppress_notifications: bool = True
    allow_critical_only: bool = False
    reason: str | None = None
    started_at: datetime | None = None
    ended_at: datetime | None = None
    policy_snapshot_json: dict[str, Any] = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class ResidentPresenceEventRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    companion_id: uuid.UUID
    realtime_session_id: uuid.UUID | None = None
    event_type: str
    event_status: str
    interruption_level: str
    requires_user_confirmation: bool = False
    delivery_surface: str
    event_summary: str | None = None
    policy_snapshot_json: dict[str, Any] = Field(default_factory=dict)
    occurred_at: datetime | None = None

    model_config = {"from_attributes": True}


class ScopedHardStopEventRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    hard_stop_scope: str
    hard_stop_status: str
    initiated_by: str
    realtime_session_id: uuid.UUID | None = None
    channel_id: uuid.UUID | None = None
    companion_id: uuid.UUID | None = None
    context_event_id: uuid.UUID | None = None
    stop_reason: str | None = None
    stops_listening: bool = True
    stops_speaking: bool = True
    stops_observing: bool = True
    stops_memory_capture: bool = True
    stops_context_capture: bool = True
    requires_audit: bool = True
    policy_snapshot_json: dict[str, Any] = Field(default_factory=dict)
    released_at: datetime | None = None

    model_config = {"from_attributes": True}


class HardStopAuditEventRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    hard_stop_event_id: uuid.UUID
    audit_event_type: str
    audit_status: str
    affected_scope: str
    affected_participant_id: uuid.UUID | None = None
    audit_summary: str | None = None
    audit_payload_json: dict[str, Any] = Field(default_factory=dict)
    occurred_at: datetime | None = None

    model_config = {"from_attributes": True}


class ScopedHardStopEventCreate(BaseModel):
    hard_stop_scope: str
    realtime_session_id: uuid.UUID | None = None
    channel_id: uuid.UUID | None = None
    companion_id: uuid.UUID | None = None
    context_event_id: uuid.UUID | None = None
    stop_reason: str | None = None
    requires_audit: bool = True

File: apps/web/lib/api/residentPresence.ts
import { apiPost } from "./client";
import type { CoPresenceInvitationRecord, MeaningfulSilenceResult, PresenceBudgetEvaluation, ResidentStatusRecord } from "@/lib/types";

function normalizeStatus(data: Partial<ResidentStatusRecord> | null): ResidentStatusRecord {
  const item = data ?? {};
  return {
    id: item.id ?? "",
    user_id: item.user_id ?? "",
    companion_id: item.companion_id ?? "",
    realtime_session_id: item.realtime_session_id ?? null,
    status_type: item.status_type ?? "available",
    status_source: item.status_source ?? "user",
    interruption_level: item.interruption_level ?? "low",
    allows_unsolicited_presence: item.allows_unsolicited_presence ?? false,
    presence_summary: item.presence_summary ?? null,
    policy_snapshot_json: item.policy_snapshot_json ?? {},
    occurred_at: item.occurred_at ?? null,
  };
}

function normalizeBudget(data: Partial<PresenceBudgetEvaluation> | null): PresenceBudgetEvaluation {
  const item = data ?? {};
  return {
    id: item.id ?? "",
    user_id: item.user_id ?? "",
    companion_id: item.companion_id ?? "",
    budget_scope: item.budget_scope ?? "day",
    budget_status: item.budget_status ?? "active",
    enforcement_policy: item.enforcement_policy ?? "queue_when_exhausted",
    max_presence_minutes: item.max_presence_minutes ?? 0,
    used_presence_minutes: item.used_presence_minutes ?? 0,
    max_interruptions: item.max_interruptions ?? 0,
    used_interruptions: item.used_interruptions ?? 0,
    budget_policy_json: item.budget_policy_json ?? {},
    decision: item.decision ?? "allowed",
    allowed: item.allowed ?? false,
  };
}

function normalizeInvitation(data: Partial<CoPresenceInvitationRecord> | null): CoPresenceInvitationRecord {
  const item = data ?? {};
  return {
    id: item.id ?? "",
    user_id: item.user_id ?? "",
    realtime_session_id: item.realtime_session_id ?? null,
    inviter_companion_id: item.inviter_companion_id ?? null,
    target_companion_id: item.target_companion_id ?? null,
    invitation_status: item.invitation_status ?? "queued",
    invitation_source: item.invitation_source ?? "user_request",
    requires_user_approval: item.requires_user_approval ?? true,
    auto_join_allowed: item.auto_join_allowed ?? false,
    memory_candidate_allowed: item.memory_candidate_allowed ?? false,
    invitation_reason: item.invitation_reason ?? null,
    policy_snapshot_json: item.policy_snapshot_json ?? {},
    expires_at: item.expires_at ?? null,
  };
}

export function setResidentStatus(data: Record<string, unknown>) {
  return apiPost<ResidentStatusRecord>("/resident-presence/status", data).then(normalizeStatus);
}

export function evaluateResidentPresenceBudget(data: Record<string, unknown>) {
  return apiPost<PresenceBudgetEvaluation>("/resident-presence/budget/evaluate", data).then(normalizeBudget);
}

export function createResidentCoPresenceInvitation(data: Record<string, unknown>) {
  return apiPost<CoPresenceInvitationRecord>("/resident-presence/invitations", data).then(normalizeInvitation);
}

export function applyMeaningfulSilence(data: Record<string, unknown>) {
  return apiPost<MeaningfulSilenceResult>("/resident-presence/meaningful-silence", data).then((result) => ({
    quiet: result?.quiet ?? {},
    focus: result?.focus ?? {},
  }));
}
